l1/l1.go

```diff
@@ -0,0 +1,111 @@
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/l1/contract"
+	"github.com/NethermindEth/juno/service"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+var ConfirmationPeriod uint64 = 16
+
+//go:generate mockgen -destination=../mocks/mock_subscriber.go -package=mocks github.com/NethermindEth/juno/l1 Subscriber
+type Subscriber interface {
+	WatchLogStateUpdate(ctx context.Context, sink chan<- *contract.StarknetLogStateUpdate) (event.Subscription, error)
+	WatchHeader(ctx context.Context, ch chan<- *types.Header) (event.Subscription, error)
+}
+
+type Client struct {
+	l1      Subscriber
+	l2Chain *blockchain.Blockchain
+	log     utils.SimpleLogger
+	q       map[uint64]*core.L1Update
+}
+
+var _ service.Service = (*Client)(nil)
+
+func NewClient(l1 Subscriber, chain *blockchain.Blockchain, log utils.SimpleLogger) *Client {
+	return &Client{
+		l1:      l1,
+		l2Chain: chain,
+		log:     log,
+		q:       make(map[uint64]*core.L1Update),
+	}
+}
+
+func (c *Client) Run(ctx context.Context) error {
+	subscribeCtx, subscribeCtxCancel := context.WithCancel(ctx)
+	defer subscribeCtxCancel()
+	buffer := 128
+
+	headerChan := make(chan *types.Header, buffer)
+	subHeader, err := c.l1.WatchHeader(subscribeCtx, headerChan)
+	if err != nil {
+		return fmt.Errorf("subscribe to L1 headers: %w", err)
+	}
+	defer subHeader.Unsubscribe()
+
+	logStateUpdateChan := make(chan *contract.StarknetLogStateUpdate, buffer)
+	subLogStateUpdate, err := c.l1.WatchLogStateUpdate(subscribeCtx, logStateUpdateChan)
+	if err != nil {
+		return fmt.Errorf("subscribe to L1 state updates: %w", err)
+	}
+	defer subLogStateUpdate.Unsubscribe()
+
+	for {
+		select {
+		case err := <-subHeader.Err():
+			return err
+		case <-ctx.Done():
+			return nil
+		case header := <-headerChan:
+		Outer:
+			for {
+				select {
+				case err := <-subLogStateUpdate.Err():
+					return err
+				case logStateUpdate := <-logStateUpdateChan:
+					if logStateUpdate.Raw.Removed {
+						for blockNumber := range c.q {
+							if blockNumber >= logStateUpdate.Raw.BlockNumber {
+								delete(c.q, blockNumber)
+							}
+						}
+						continue
+					}
+					c.q[logStateUpdate.Raw.BlockNumber] = &core.L1Update{
+						BlockNumber: logStateUpdate.BlockNumber.Uint64(),
+						StateRoot:   new(felt.Felt).SetBigInt(logStateUpdate.GlobalRoot),
+					}
+				default:
+					break Outer
+				}
+			}
+			if err := c.updateBlockchain(header.Number.Uint64()); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (c *Client) updateBlockchain(l1BlockNumber uint64) error {
+	for blockNumber, update := range c.q {
+		// Wait for at least one block just in case the tip of the L1 chain is reorged.
+		if blockNumber > l1BlockNumber-ConfirmationPeriod {
+			continue
+		}
+		if err := c.l2Chain.SetL1Update(update); err != nil {
```

Comment:
we wont be calling `SetL1Update` with ordered `blockNumber`s now.
`SetL1Update` should make sure the incoming update is more recent.

Reply:
Dont we still have the same issue here?

when we `range c.q`, we will loop over the `blockNumber`s in random order